File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-settings/src/state/settingsState.ts
import type { SettingsPageResult } from "../types/settingsModels.js";

export interface SettingsState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: SettingsPageResult<unknown>;
  readonly error?: string;
}

export function initialSettingsState(): SettingsState {
  return { status: "idle" };
}

export function reduceSettingsState(
  state: SettingsState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: SettingsPageResult<unknown> }
    | { type: "failed"; error: string },
): SettingsState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}

File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_settings/lib/src/state/settings_state.dart
import '../models/settings_models.dart';

enum SettingsStatus { idle, loading, ready, error }

class SettingsState {
  const SettingsState({
    this.status = SettingsStatus.idle,
    this.lastResult,
    this.error,
  });

  final SettingsStatus status;
  final SettingsPageResult<Object?>? lastResult;
  final String? error;

  SettingsState loading() => const SettingsState(status: SettingsStatus.loading);

  SettingsState loaded(SettingsPageResult<Object?> result) =>
      SettingsState(status: SettingsStatus.ready, lastResult: result);

  SettingsState failed(String message) =>
      SettingsState(status: SettingsStatus.error, error: message);
}
